fairlearn/metrics/metrics_engine.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import numpy as np
+
+from . import GroupMetricResult
+
+_MESSAGE_NON_BINARY = "Array {0} contains values other than 0 and 1"
+_MESSAGE_SIZE_MISMATCH = "Array {0} is not the same size as {1}"
+
+
+def metric_by_groups(metric_function, y_true, y_pred, group_data, sample_weight=None):
+    """ Applies a metric to each subgroup of a set of data
+
+    :param metric_function
+    :type Function with signature (y_true, y_pred, sample_weight=None) which returns a scalar
+
+    :param y_true
+    :type Array of actual results (must be 0 or 1)
+
+    :param y_pred
+    :type Array of predicted results (must be 0 or 1)
+
+    :param group_data
+    :type Array indicating the group to which each result belongs
+
+    :param sample_weight
+    :type Array of weights to apply to each result
+    """
+    _check_binary(y_true, "y_true")
+    _check_binary(y_pred, "y_pred")
+    _check_array_sizes(y_true, y_pred, 'y_true', 'y_pred')
+    _check_array_sizes(y_true, group_data, 'y_true', 'group_data')
+    if sample_weight is not None:
+        _check_array_sizes(y_true, sample_weight, 'y_true', 'sample_weight')
+
+    result = GroupMetricResult()
+
+    groups = np.unique(group_data)
+
+    result.metric = metric_function(y_true, y_pred, sample_weight=sample_weight)
+
+    # The slicing we use requires Numpy arrays
+    y_a = np.asarray(y_true)
+    y_p = np.asarray(y_pred)
+    g_d = np.asarray(group_data)
+    s_w = None
+    if sample_weight is not None:
+        s_w = np.asarray(sample_weight)
+
+    for group in groups:
+        group_indices = (group == g_d)
+        group_actual = y_a[group_indices]
+        group_predict = y_p[group_indices]
+        group_weight = None
+        if s_w is not None:
+            group_weight = s_w[group_indices]
+        result.group_metrics[group] = metric_function(group_actual,
+                                                      group_predict,
+                                                      sample_weight=group_weight)
```

Comment:
my suggestion is to do an explicit if, and only include sample_weight in the call to metric_function when it's not None. this is to allow disparity for metrics that do not support sample_weight.

Reply:
Done